Improve EFS and EBS boto3 calls to reduce pcluster CLI/API running time

hanwen-pcluste opened the pull request:
Using "Hide Whitespace" to review this PR is helpful

1. Expand EFS mount targets validator to check both head subnet and compute subnets availability zones. Previously, only head node subnet is checked.

2. Add custom caching for boto3 calls that take lists as arguments. The standard @Cache utility is not compatible with functions with lists arguments

3. Collapse some FSx validators to make single calls for multiple file systems

4. There are no good EFS boto3 calls to do similar collapse as FSx

Signed-off-by: Hanwen <hanwenli@amazon.com>

### Tests
* all tests in configs/pcluster3.yaml have been passed

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

enrico-usai commented on cli/src/pcluster/aws/ec2.py:
```diff
@@ -67,7 +71,20 @@ def get_default_instance_type(self):
     def describe_subnets(self, subnet_ids):
         """Return a list of subnets."""
-        return list(self._paginate_results(self._client.describe_subnets, SubnetIds=subnet_ids))
+        result = []
```
If we're going to add this extra behaviour we need to write uni test for this method.

hanwen-pcluste replied:
Added

enrico-usai commented on cli/src/pcluster/aws/ec2.py:
```diff
@@ -34,6 +34,10 @@ class Ec2Client(Boto3Client):
+        self.security_groups_cache = {}
+        Cache.add_cache(self.security_groups_cache)
+        self.subnets_cache = {}
+        Cache.add_cache(self.subnets_cache)
```
I'd add a comment here too, describing why we are adding these two dict into the cache.

enrico-usai replied:
minor: I'd define all the default first, then I'd add them to the cache.
```python

        self.security_groups_cache = {}
        self.subnets_cache = {}
        Cache.add_cache(self.security_groups_cache)
        Cache.add_cache(self.subnets_cache)
```

hanwen-pcluste replied:
Not valid anymore

enrico-usai commented on cli/src/pcluster/aws/ec2.py:
```diff
@@ -381,18 +398,33 @@ def get_ebs_snapshot_info(self, ebs_snapshot_id):
     def describe_security_groups(self, security_group_ids):
         """Describe a single security group."""
-        return self._client.describe_security_groups(GroupIds=security_group_ids)
+        result = []
```
the comment of the function is wrong (it's not your fault).

hanwen-pcluste replied:
Fixed

enrico-usai commented on cli/src/pcluster/aws/ec2.py:
```diff
@@ -381,18 +398,33 @@ def get_ebs_snapshot_info(self, ebs_snapshot_id):
+                result.append(cached_data)
+            else:
+                missed_security_group_ids.append(security_group_id)
+        if missed_security_group_ids:
```
same here, we need unit tests to cover all these cases.

hanwen-pcluste replied:
Added

enrico-usai commented on cli/src/pcluster/validators/cluster_validators.py:
```diff
@@ -403,55 +403,76 @@ def _check_sg_rules_for_port(rule, port_to_check):
+            for file_system in file_systems:
+                # Check to see if fs is in the same VPC as the stack
+                file_system_data = file_system.file_system_data
+                file_system_id = file_system_data.get("FileSystemId")
```
We shouldn't retrieve FileSystemId, VpcId and other file system data fields from here.
We have the `FsxFileSystemInfo` class that is useful to hide the internal implementation. 

I know some of them were already here but let's improve the code by adding `id`, `vpc_id`, `network_interface_ids` and other useful methods in the `FsxFileSystemInfo` class and use them here.

In this way the validator doesn't need to know the boto3 implementation and internal boto3 keys. This avoids typos and provides better layer separation.

enrico-usai replied:
`file_system_data` should be a private attribute of the `FsxFileSystemInfo` class.

hanwen-pcluste replied:
Fixed

enrico-usai commented on cli/src/pcluster/validators/cluster_validators.py:
```diff
@@ -567,22 +588,22 @@ class EfsIdValidator(Validator):  # TODO add tests
-                    ),
-                    FailureLevel.ERROR,
-                )
+    def _validate(self, efs_id, avail_zones: set, are_all_security_groups_customized):
```
I know it's not part of your change but we still don't have unit tests for this validator.

enrico-usai commented on cli/src/pcluster/aws/common.py:
```diff
@@ -184,6 +184,12 @@ def clear_all():
             cache.clear()
 
+    @staticmethod
+    def add_cache(custom_cache):
```
let's add typing to clarify which kind of items we're adding.

hanwen-pcluste replied:
After chatting with demartinofra, we moved away from this logic. Then we added a `reset() in AWSApi class to reset the singleton to None when caches need to be cleared

enrico-usai commented on cli/src/pcluster/aws/common.py:
```diff
@@ -184,6 +184,12 @@ def clear_all():
+    @staticmethod
+    def add_cache(custom_cache):
+        """Add custom cache, so the cache created outside this wrapper can be cleared to."""
+        # Todo: Validate this approach
```
todo?

hanwen-pcluste replied:
Not valid anymore

I also validate our new approach today

enrico-usai commented on cli/src/pcluster/aws/fsx.py:
```diff
@@ -17,17 +17,32 @@ class FSxClient(Boto3Client):
     def __init__(self):
         super().__init__("fsx")
+        self.cache = {}
+        Cache.add_cache(self.cache)
```
`Cache.add_cache(self.cache)` ? :D

There is something we can improve in both variable names and methods.
1. let's use a more self-explicative name for the variable `file_systems_cache` to be clear what we're going to cache in this variable.
2. what about renaming `add_cache` to: `insert` ?

hanwen-pcluste replied:
Hahaha. Not valid anymore

enrico-usai commented on cli/src/pcluster/aws/fsx.py:
```diff
@@ -17,17 +17,32 @@ class FSxClient(Boto3Client):
+        missed_fsx_fs_ids = []
+        for file_system_id in fsx_fs_ids:
+            cached_data = self.cache.get(file_system_id)
+            if cached_data:
```
same here we need unit tests for this custom logic.

hanwen-pcluste replied:
I will add an unit test tomorrow if the other new unit tests look good to you

enrico-usai replied:
Unit tests for other caches looks good to me, thanks for creating them!

hanwen-pcluste replied:
Done

hanwen-cluster committed:
Improve EFS and EBS boto3 calls to reduce pcluster CLI/API running time

Using "Hide Whitespace" to review this commit is helpful

1. Expand EFS mount targets validator to check both head subnet and compute subnets availability zones. Previously, only head node subnet is checked.

2. Add custom caching for boto3 calls that take lists as arguments. The standard @Cache utility is not compatible with functions with lists arguments

3. Collapse some FSx validators to make single calls for multiple file systems

4. There are no good EFS boto3 calls to do similar collapse as FSx

Signed-off-by: Hanwen <hanwenli@amazon.com>

enrico-usai commented on cli/tests/pcluster/aws/test_ec2.py:
```diff
@@ -331,3 +332,77 @@ def test_run_instances_dryrun(boto3_stubber, error_code, raise_exception):
+    # First boto3 call. Nothing has been cached
+    security_group = "sg-123"
+    additional_security_group = "sg-234"
+    # The first mocked request and the third are about the same security group. However, the ip permission of
```
missing something in this comment. "However, the ip permission of the SG change from empty to {}".

hanwen-pcluste replied:
Done

enrico-usai commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -1228,6 +1230,15 @@ def compute_subnet_ids(self):
+    def availability_zones(self):
+        """Return the list of all availability zones in the cluster."""
+        result = set(self.head_node.networking.availability_zone)
+        for queue in self.scheduling.queues:
```
What about using `compute_subnet_ids` that is an already filtered set?

hanwen-pcluste replied:
Done

hanwen-pcluste committed:
Merge branch 'develop' into developApr12
